<a href="https://colab.research.google.com/github/Aleena24/DL_lab/blob/main/lab3_A_optimization.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

given revenue function
ƒ (w) = 0.5w2-30w +100

*   w = price of product
*   0.5w2 represents the revenue generated from selling the product
*   −30w represents the cost associated with marketing/advertising the product
*   100 represents a constant term that could be interpreted as fixed costs or
    other constant factors affecting revenue.





optimal price = w which minimizes the reevnue lost.

revenue function = -revenue lost

minimizing revenue lost = maximizing revenue

optimization process uses momentum based process to adjust the price based on the gradient





In [22]:
def calculate_revenue_loss(price):
    return 0.5 * price**2 - 30 * price + 100

def calculate_gradient(price):
    return price - 30

def optimize_price(initial_price, learning_rate, momentum, num_iterations):
    price = initial_price
    velocity = 0

    for i in range(num_iterations):
        gradient = calculate_gradient(price)
        velocity = momentum * velocity + learning_rate * gradient
        price -= velocity

        revenue_loss = calculate_revenue_loss(price)
        print(f"Iteration {i+1}: Price = {price:.2f}, Revenue Loss = {revenue_loss:.2f}")

    return price

# Parameters
initial_price = 20
learning_rate = 0.1
momentum = 0.9
num_iterations = 3

# Perform momentum-based optimization
optimal_price = optimize_price(initial_price, learning_rate, momentum, num_iterations)
print(f"\nOptimal Price: {optimal_price:.2f}")


Iteration 1: Price = 21.00, Revenue Loss = -309.50
Iteration 2: Price = 22.80, Revenue Loss = -324.08
Iteration 3: Price = 25.14, Revenue Loss = -338.19

Optimal Price: 25.14


The optimal price obtained after the optimization process is 25.14

This price minimizes the revenue loss associated with marketing/advertisements for the product, as determined by the optimization algorithm.